test/helpers/position_assertions.cc

```diff
@@ -1232,6 +1233,80 @@ string HoverAssertion::toString() const {
+        fmt::format(
+            "Improperly formatted hover-multiline assertion. Expected '<lineno> <contents>'. Found '{}'",
+            assertionContents, filename));
+
+    return nullptr;
+}
+HoverMultilineAssertion::HoverMultilineAssertion(string_view filename, unique_ptr<Range> &range, int assertionLine,
+                                                 int lineno, string_view message)
+    : RangeAssertion(filename, range, assertionLine), lineno(lineno), message(string(message)) {}
+
+void HoverMultilineAssertion::checkAll(const vector<shared_ptr<RangeAssertion>> &assertions,
+                              const UnorderedMap<string, shared_ptr<core::File>> &sourceFileContents,
+                              LSPWrapper &wrapper, int &nextId, string errorPrefix) {
+    for (auto assertion : assertions) {
+        if (auto assertionOfType = dynamic_pointer_cast<HoverMultilineAssertion>(assertion)) {
+            assertionOfType->check(sourceFileContents, wrapper, nextId, errorPrefix);
+        }
+    }
+}
+
+// Returns `true` if `line` matches a full line of text in `text`.
+bool containsLineAt(string_view text, string_view line, int lineno) {
+    vector<string> lines = absl::StrSplit(text, "\n");
+    return lineno < lines.size() && lines[lineno] == line;
+}
+
+void HoverMultilineAssertion::check(const UnorderedMap<string, shared_ptr<core::File>> &sourceFileContents, LSPWrapper &wrapper,
+                           int &nextId, string errorPrefix) {
+    const auto &config = wrapper.config();
+    auto uri = filePathToUri(config, filename);
+    auto pos = make_unique<TextDocumentPositionParams>(make_unique<TextDocumentIdentifier>(uri), range->start->copy());
+    auto id = nextId++;
+    auto msg = make_unique<LSPMessage>(make_unique<RequestMessage>("2.0", id, LSPMethod::TextDocumentHover, move(pos)));
+    auto responses = getLSPResponsesFor(wrapper, move(msg));
+    REQUIRE_EQ(responses.size(), 1);
+    auto &responseMsg = responses.at(0);
+    REQUIRE(responseMsg->isResponse());
+    auto &response = responseMsg->asResponse();
+    REQUIRE_MESSAGE(response.result.has_value(), response.error.value()->message);
+    auto &hoverResponse = get<variant<JSONNullObject, unique_ptr<Hover>>>(*response.result);
+    auto hoverContents = hoverToString(hoverResponse);
+
+    // Match a full line. Makes it possible to disambiguate `String` and `T.nilable(String)`.
+    if (!containsLineAt(hoverContents, this->message, this->lineno)) {
+        auto sourceLine = getSourceLine(sourceFileContents, filename, range->start->line);
+        ADD_FAIL_CHECK_AT(
+            filename.c_str(), range->start->line + 1,
+            fmt::format("{}Expected hover contents:\n{}\nFound hover contents:\n{}", errorPrefix,
+                        prettyPrintRangeComment(sourceLine, *range, toString()),
+                        prettyPrintRangeComment(sourceLine, *range,
+                            fmt::format("hover-multiline: {}", hoverContents))));
+    }
```

Comment:
I'm unsure of this failure, currently it looks like this: 
```
===============================================================================
test/lsp_test_runner.cc:525:
TEST CASE:  LSPTest

test/testdata/lsp/hover.rb:83: ERROR: Expected hover contents:
83:   def tests_return_markdown
          ^^^^^^^^^^^^^^^^^^^^^ hover-multiline: 12 Tests return markdown output
Found hover contents:
83:   def tests_return_markdown
          ^^^^^^^^^^^^^^^^^^^^^ hover-multiline: ```ruby
sig {void}
def tests_return_markdown; end
```

---

Tests return markdown output

===============================================================================
```

I'm wondering if in targeting the specific line, it makes more sense to have `<lineno> <contents>` where `contents` is the actual value of that expected line or empty.

Reply:
Thinking about it, it's helpful to be able to see the complete output on test failure.

Reply:
What about "Found" -> "Complete" and "Expected hover contents" -> "Expected line N of hover contents"